prow/plugins/project/project.go

```
@@ -0,0 +1,348 @@
+
+	// No project is provided
+	if len(matches) == 0 {
+		msg := invalidNumArgs
+		return "", "", false, msg
+	}
+	proposedProject := matches[1]
+	proposedColumnName := ""
+	if len(matches) > 1 && proposedProject != clearKeyword {
+		proposedColumnName = matches[2]
+	}
+	// If command is to clear and the project is provided
+	if proposedProject == clearKeyword {
+		if len(matches) > 2 && matches[2] != "" {
+			proposedProject = matches[2]
+			shouldClear = true
+		} else {
+			proposedProject = ""
+		}
+	}
+
+	return proposedProject, proposedColumnName, shouldClear, ""
+}
+
+func handle(gc githubClient, log *logrus.Entry, e *github.GenericCommentEvent, projectConfig plugins.ProjectConfig) error {
+	if e.Action != github.GenericCommentActionCreated {
+		return nil
+	}
+
+	org := e.Repo.Owner.Login
+	repo := e.Repo.Name
+	proposedProject := ""
+	proposedColumnName := ""
+	shouldClear := false
+	msg := ""
+	if proposedProject, proposedColumnName, shouldClear, msg = parseCommand(e.Body); proposedProject == "" {
+		return gc.CreateComment(org, repo, e.Number, plugins.FormatResponseRaw(e.Body, e.HTMLURL, e.User.Login, msg))
+	}
+
+	isAMember := false
+	var err error
+	if maintainerTeamID, maintainerTeamName := getMaintainerTeam(projectConfig, org, repo); maintainerTeamName != "" && maintainerTeamID != -1 {
+		// See if the current user is part of the team in order to set the project/column
+		if isAMember, err = gc.TeamHasMember(maintainerTeamID, e.User.Login); err != nil {
+			return err
+		}
+	}
+	if !isAMember {
+		// not in the project maintainers team
+		msg = fmt.Sprintf(notATeamMemberMsg, org, repo, org, repo)
+		return gc.CreateComment(org, repo, e.Number, plugins.FormatResponseRaw(e.Body, e.HTMLURL, e.User.Login, msg))
+	}
+
+	repos, err := gc.GetRepos(org, false)
+	if err != nil {
+		msg = fmt.Sprintf(errListingReposMsg, org)
+		return gc.CreateComment(org, repo, e.Number, plugins.FormatResponseRaw(e.Body, e.HTMLURL, e.User.Login, msg))
```

So some of these errors are useful to users -- like if they try to use this command with a project that doesn't exist, or if they're not in the group. However, this call will fail only if there is an infrastructure error -- if we don't have permissions or can't hit the API or something. We should just return the error here